test/jasmine/tests/plot_api_test.js

```diff
@@ -3027,70 +3219,127 @@ describe('Test plot api', function() {
-                                // as a hack to tell Plotly.react to ignore changes.
-                                // Can we kill this?
-                                delete ax.range;
-                            }
-                        }
-                    });
-                    out.data.forEach(function(trace) {
-                        if(trace.type === 'contourcarpet') {
-                            delete trace.zmin;
-                            delete trace.zmax;
-                        }
-                    });
+                    }
+                });
+                out.data.forEach(function(trace) {
+                    if(trace.type === 'contourcarpet') {
+                        delete trace.zmin;
+                        delete trace.zmax;
+                    }
+                });
+
+                return out;
+            }
 
-                    return out;
+            // Make sure we define `_length` in every trace *in supplyDefaults*.
+            // This is only relevant for traces that *have* a 1D concept of length,
+            // and in addition to simplifying calc/plot logic later on, ths serves
+            // as a signal to transforms about how they should operate. For traces
+            // that do NOT have a 1D length, `_length` should be `null`.
+            var mockGD = Lib.extendDeep({}, mock);
+            supplyAllDefaults(mockGD);
+            expect(mockGD._fullData.length).not.toBeLessThan((mock.data || []).length, mockSpec[0]);
+            mockGD._fullData.forEach(function(trace, i) {
+                var len = trace._length;
+                if(trace.visible !== false && len !== null) {
+                    expect(typeof len).toBe('number', mockSpec[0] + ' trace ' + i + ': type=' + trace.type);
                 }
 
-                Plotly.newPlot(gd, mock)
-                .then(countPlots)
-                .then(function() {
-                    initialJson = fullJson();
-                    return Plotly.react(gd, mock);
-                })
-                .then(function() {
-                    expect(fullJson()).toEqual(initialJson);
-                    countCalls({});
-                })
-                .catch(failTest)
-                .then(done);
+                typesTested[trace.type]++;
+
+                if(trace.transforms) {
+                    trace.transforms.forEach(function(transform) {
+                        typesTested[transform.type]++;
+                    });
+                }
             });
+
+            Plotly.newPlot(gd, mock)
+            .then(countPlots)
+            .then(function() {
+                initialJson = fullJson();
+
+                return Plotly.react(gd, mock);
+            })
+            .then(function() {
+                expect(fullJson()).toEqual(initialJson);
+                countCalls({});
+            })
+            .catch(failTest)
+            .then(done);
+        }
+
+        mockList.forEach(function(mockSpec) {
```

Comment:
Is there anything we can think of that could be asserted about the cartesian product here? I.e. given `M1` and `M2` is there anything in the JSON representation that we want to try to assert about `M1.reactWith(M2).reactWith(M1)` ?

Reply:
There are likely some things that are desirable that are already true in all cases and might be worth locking down, and some things that are desirable that are true in all but a small subset of cases and could be fixed easily (low-hanging fruits). There are also probably a large number of desirable things that aren't true and that are hard to fix, which I'm less interested in in the short run.

Reply:
I tested a few of those transitions in https://github.com/plotly/plotly.js/pull/2577/commits/fcc459d985f0a3a80826da8df51b2daf150a4634#diff-5a140ce1cc87a420496039bbf8d699f1R3114 - and in fact after this PR (knock wood) I kind of don't think there will be a lot of nasty issues uncovered by testing this cartesian product. It's also going to be a very long test if you want to cover every combination, so it sounds to me more like we'd want to test them all once, offline, find the subset that fail, and incorporate just those into a new test.